jax/lax/lax_parallel.py

```diff
@@ -601,13 +607,19 @@ def all_gather(x, axis_name):
     x: array(s) with a mapped axis named ``axis_name``.
     axis_name: hashable Python object used to name a pmapped axis (see the
       :func:`jax.pmap` documentation for more details).
+    axis_index_groups: optional list of lists containing axis indices (e.g. for
+      an axis of size 4, [[0, 1], [2, 3]] would run all gather over the first
+      two and last two replicas). Groups must cover all axis indices exactly
+      once, and all groups must be the same size.
+    axis_index_to_group_index: list of indices mapping from global axis index
```

Comment:
Since`axis_index_to_group_index` is uniquely determined by `axis_index_groups`, I feel that it is redundant to ask the user to provide the argument. I think that a better approach is to add code to compute it. I think it makes sense to have this code run at tracing time (use `np` instead of `jnp` for the calculations).

Reply:
Done.